### DISCUSSION TITLE
Placement(XYZ) of ifcCurtainwall and ifcSpace are zeros

### DISCUSSION THREAD

**Discussion by ColaZhang22** [IfcOpenShell/IfcOpenShell#2625] [Q&A] [answered]
Hello,
I am newer to ifc file and try to extract all coordinates about elements from ifcfile by python and ifcopenshell. 
I successfully extract placement of walls, doors and columns. However, when I try to extract placement of curtainwall and space, return matrix is always zero.
I try **ifcopenshell.util.element.get_psets(space)** and 
        **ifcopenshell.util.placement.get_local_placement(space.ObjectPlacement)**  these two methods but no work. And I search back along PlacementRelto and find all they are zeros. Thus, I do not know the reason about these failure.
So I want to know if I make some mistakes in process of using these method or have other methods to get coordinates of IfcSpace.
Thanks 
![image](https://user-images.githubusercontent.com/118124550/204577433-3ae5edee-89dd-47ee-b9a9-3a573e0b9f68.png)

**Comment by Moult** [comment, marked as answer]
Which software created your model? If it works for doors, but not for spaces, it may be a bug in the software that produced the IFC - I have seen software before (such as Revit) that have simply used a placement of 0,0,0 for many shapes. It's pretty annoying, I'd agree.

**Reply by ColaZhang22** [reply]
Thanks for your reply.   
Actually Revit created my model and I exported this model into an IFC file. 
My target is to extract placement information of IFC and build an environment for Reinforcement learning.
Thus, could you recommend a stable software which be able to build correct model?
Thanks again.

**Reply by Moult** [reply]
One option is to fix the placements with IfcOpenShell - you can use the geometry iterator to fetch the individual vertices of the spaces, and then calculate a centroid or min/max: https://blenderbim.org/docs-python/ifcopenshell-python/geometry_processing.html then modify the space geometry and placement accordingly.

Maybe give ArchiCAD a shot? Or maybe use some open source software like the BlenderBIM Add-on and/or FreeCAD? The IFC support in the BlenderBIM Add-on is pretty advanced relative to proprietary systems and you won't run into "export" issues since we work with IFC natively.

**Reply by ColaZhang22** [reply]
Thanks for your recommendation and advice, it is really useful and solve an annoying question.
I would try your methods. Thanks!

**Reply by aothms** [reply]
IfcCurtainWall also doesn't have  a geometric representation, but is rather a decomposition of members and plates. Being some sort of a system/grouping entity in revit I guess they just don't maintain a placement. But you could take the minimum (whatever that might mean) placement of its decomposing members. For space I'm a bit surprised.
